### ISSUE TITLE
Question: msk server in port 443

### ISSUE THREAD

**Issue by rajk0007** [grepplabs/kafka-proxy#32]
My requirement is expose msk in public internet on port 443. I am trying to run this proxy in eks fronting MSK . 

```
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app: kafka-proxy-1
  name: kafka-proxy-1
spec:
  externalTrafficPolicy: Cluster
  ports:
  - name: kafka-proxy-1
    port: 443
    protocol: TCP
    targetPort: 443
  selector:
    app: kafka-proxy-1
  sessionAffinity: None
  type: LoadBalancer
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app: kafka-proxy-2
  name: kafka-proxy-2
spec:
  externalTrafficPolicy: Cluster
  ports:
  - name: kafka-proxy-2
    port: 443
    protocol: TCP
    targetPort: 443
  selector:
    app: kafka-proxy-2
  sessionAffinity: None
  type: LoadBalancer
---

apiVersion: v1
kind: Service
metadata:
  labels:
    app: kafka-proxy-3
  name: kafka-proxy-3
spec:
  externalTrafficPolicy: Cluster
  ports:
  - name: kafka-proxy-3
    port: 443
    protocol: TCP
    targetPort: 443
  selector:
    app: kafka-proxy-3
  sessionAffinity: None
  type: LoadBalancer

---
apiVersion: apps/v1
kind: Deployment
metadata:
   name: kafka-proxy-1
spec:
  replicas: 1
  selector:
    matchLabels:
      app: kafka-proxy-1
  template:
    metadata:
      labels:
        app: kafka-proxy-1
    spec:
      containers:
        - name: kafka-proxy
          securityContext:
            capabilities:
              add: ["NET_ADMIN", "SYS_TIME","NET_BIND_SERVICE"]
          image: grepplabs/kafka-proxy:latest
          args:
            - 'server'
            - '--log-format=json'
            - '--bootstrap-server-mapping=b-1.pdp-kafka.h5u916.c3.kafka.us-east-1.amazonaws.com:9094,0.0.0.0:443,a572bcec9048311ea874d02324e2a15c-102380383.us-east-1.elb.amazonaws.com:443'
            - '--external-server-mapping=b-1.pdp-kafka.h5u916.c3.kafka.us-east-1.amazonaws.com:9094,a572bcec9048311ea874d02324e2a15c-102380383.us-east-1.elb.amazonaws.com:443'
            - '--external-server-mapping=b-2.pdp-kafka.h5u916.c3.kafka.us-east-1.amazonaws.com:9094,a573b1e97048311ea874d02324e2a15c-2011622944.us-east-1.elb.amazonaws.com:443'
            - '--external-server-mapping=b-3.pdp-kafka.h5u916.c3.kafka.us-east-1.amazonaws.com:9094,a574a0444048311ea874d02324e2a15c-972090528.us-east-1.elb.amazonaws.com:443'
            - '--tls-enable'
            - '--log-level=debug'
            - '--dynamic-listeners-disable'
            - '--tls-insecure-skip-verify'
            - '--proxy-request-buffer-size=32768'
            - '--proxy-response-buffer-size=32768'
            - '--proxy-listener-read-buffer-size=32768'
            - '--proxy-listener-write-buffer-size=131072'
            - '--kafka-connection-read-buffer-size=131072'
            - '--kafka-connection-write-buffer-size=32768'
          ports:
          - name: kafka-port1
            containerPort: 443


---
apiVersion: apps/v1
kind: Deployment
metadata:
   name: kafka-proxy-2
spec:
  replicas: 1
  selector:
    matchLabels:
      app: kafka-proxy-2
  template:
    metadata:
      labels:
        app: kafka-proxy-2
    spec:
      containers:
        - name: kafka-proxy
          securityContext:
            capabilities:
              add: ["NET_ADMIN", "SYS_TIME","NET_BIND_SERVICE"]
          image: grepplabs/kafka-proxy:latest
          args:
            - 'server'
            - '--log-format=json'
            - '--bootstrap-server-mapping=b-2.pdp-kafka.h5u916.c3.kafka.us-east-1.amazonaws.com:9094,0.0.0.0:443,a573b1e97048311ea874d02324e2a15c-2011622944.us-east-1.elb.amazonaws.com:443'
            - '--external-server-mapping=b-1.pdp-kafka.h5u916.c3.kafka.us-east-1.amazonaws.com:9094,a572bcec9048311ea874d02324e2a15c-102380383.us-east-1.elb.amazonaws.com:443'
            - '--external-server-mapping=b-2.pdp-kafka.h5u916.c3.kafka.us-east-1.amazonaws.com:9094,a573b1e97048311ea874d02324e2a15c-2011622944.us-east-1.elb.amazonaws.com:443'
            - '--external-server-mapping=b-3.pdp-kafka.h5u916.c3.kafka.us-east-1.amazonaws.com:9094,a574a0444048311ea874d02324e2a15c-972090528.us-east-1.elb.amazonaws.com:443'
            - '--tls-enable'
            - '--log-level=debug'
            - '--dynamic-listeners-disable'
            - '--tls-insecure-skip-verify'
            - '--proxy-request-buffer-size=32768'
            - '--proxy-response-buffer-size=32768'
            - '--proxy-listener-read-buffer-size=32768'
            - '--proxy-listener-write-buffer-size=131072'
            - '--kafka-connection-read-buffer-size=131072'
            - '--kafka-connection-write-buffer-size=32768'
          ports:
          - name: kafka-port2
            containerPort: 443


---
apiVersion: apps/v1
kind: Deployment
metadata:
   name: kafka-proxy-3
spec:
  replicas: 1
  selector:
    matchLabels:
      app: kafka-proxy-3
  template:
    metadata:
      labels:
        app: kafka-proxy-3
    spec:
      containers:
        - name: kafka-proxy
          securityContext:
            capabilities:
              add: ["NET_ADMIN", "SYS_TIME","NET_BIND_SERVICE"]
          image: grepplabs/kafka-proxy:latest
          args:
            - 'server'
            - '--log-format=json'
            - '--external-server-mapping=b-1.pdp-kafka.h5u916.c3.kafka.us-east-1.amazonaws.com:9094,a572bcec9048311ea874d02324e2a15c-102380383.us-east-1.elb.amazonaws.com:443'
            - '--external-server-mapping=b-2.pdp-kafka.h5u916.c3.kafka.us-east-1.amazonaws.com:9094,a573b1e97048311ea874d02324e2a15c-2011622944.us-east-1.elb.amazonaws.com:443'
            - '--external-server-mapping=b-3.pdp-kafka.h5u916.c3.kafka.us-east-1.amazonaws.com:9094,a574a0444048311ea874d02324e2a15c-972090528.us-east-1.elb.amazonaws.com:443'
            - '--bootstrap-server-mapping=b-3.pdp-kafka.h5u916.c3.kafka.us-east-1.amazonaws.com:9094,0.0.0.0:443,a574a0444048311ea874d02324e2a15c-972090528.us-east-1.elb.amazonaws.com:443'
            - '--tls-enable'
            - '--log-level=debug'
            - '--dynamic-listeners-disable'
            - '--tls-insecure-skip-verify'
            - '--proxy-request-buffer-size=32768'
            - '--proxy-response-buffer-size=32768'
            - '--proxy-listener-read-buffer-size=32768'
            - '--proxy-listener-write-buffer-size=131072'
            - '--kafka-connection-read-buffer-size=131072'
            - '--kafka-connection-write-buffer-size=32768'
          ports:
          - name: kafka-port3
            containerPort: 443

```

where *.us-east-1.elb.amazonaws.com:443 are the external ip i get from eks. 

I get this error in my pod

```{"@level":"info","@message":"Starting kafka-proxy version unknown","@timestamp":"2019-11-11T22:45:59Z"}
{"@level":"info","@message":"Bootstrap server b-1.pdp-kafka.h5u916.c3.kafka.us-east-1.amazonaws.com:9094 advertised as a572bcec9048311ea874d02324e2a15c-102380383.us-east-1.elb.amazonaws.com:443","@timestamp":"2019-11-11T22:45:59Z"}
{"@level":"info","@message":"External server b-2.pdp-kafka.h5u916.c3.kafka.us-east-1.amazonaws.com:9094 advertised as a573b1e97048311ea874d02324e2a15c-2011622944.us-east-1.elb.amazonaws.com:443","@timestamp":"2019-11-11T22:45:59Z"}
{"@level":"info","@message":"External server b-3.pdp-kafka.h5u916.c3.kafka.us-east-1.amazonaws.com:9094 advertised as a574a0444048311ea874d02324e2a15c-972090528.us-east-1.elb.amazonaws.com:443","@timestamp":"2019-11-11T22:45:59Z"}
{"@level":"info","@message":"Listening on 0.0.0.0:443 ([::]:443) for remote b-1.pdp-kafka.h5u916.c3.kafka.us-east-1.amazonaws.com:9094","@timestamp":"2019-11-11T22:45:59Z"}
{"@level":"info","@message":"Ready for new connections","@timestamp":"2019-11-11T22:45:59Z"}
{"@level":"info","@message":"New connection for b-1.pdp-kafka.h5u916.c3.kafka.us-east-1.amazonaws.com:9094","@timestamp":"2019-11-11T22:46:01Z"}
{"@level":"info","@message":"Client closed local connection on 20.10.3.191:443 from 20.10.1.197:39176 (b-1.pdp-kafka.h5u916.c3.kafka.us-east-1.amazonaws.com:9094)","@timestamp":"2019-11-11T22:46:01Z"}
{"@level":"info","@message":"New connection for b-1.pdp-kafka.h5u916.c3.kafka.us-east-1.amazonaws.com:9094","@timestamp":"2019-11-11T22:46:05Z"}
{"@level":"info","@message":"Client closed local connection on 20.10.3.191:443 from 20.10.2.185:7332 (b-1.pdp-kafka.h5u916.c3.kafka.us-east-1.amazonaws.com:9094)","@timestamp":"2019-11-11T22:46:05Z"}
{"@level":"info","@message":"New connection for b-1.pdp-kafka.h5u916.c3.kafka.us-east-1.amazonaws.com:9094","@timestamp":"2019-11-11T22:46:09Z"}
{"@level":"info","@message":"Client closed local connection on 20.10.3.191:443 from 20.10.3.252:36298 (b-1.pdp-kafka.h5u916.c3.kafka.us-east-1.amazonaws.com:9094)","@timestamp":"2019-11-11T22:46:09Z"}
{"@level":"info","@message":"New connection for b-1.pdp-kafka.h5u916.c3.kafka.us-east-1.amazonaws.com:9094","@timestamp":"2019-11-11T22:46:11Z"}
{"@level":"info","@message":"Client closed local connection on 20.10.3.191:443 from 20.10.1.197:39186 (b-1.pdp-kafka.h5u916.c3.kafka.us-east-1.amazonaws.com:9094)","@timestamp":"2019-11-11T22:46:11Z"}```

**Comment by rajk0007** [comment]
Here are my pods and nodes

**Comment by rajk0007** [comment]
@everesio I tried your example [here]( https://gist.github.com/everesio/e6fae11ed69099ae3f867eddcd83db82). But that did not work .

**Comment by everesio** [comment]
Could you provide your client configuration ?
How do you want to access the proxy , from a client running in or outside kubernetes ?

**Comment by rajk0007** [comment]
@everesio  My client is outside of kubernetes cluster. 

I was finally figured out the issue. I was using nlb with the k8s service. In AWS, when nlb is created, it creates a target group which comes with tcp health check for each port exposed in service. kafka proxy didn;t understand this health check and failed with the message above. Solution was to use elb instead of nlb and expose port 9080(default healthcheck port) Doing this, there is no backend target group creation and healthcheck is only for first port listed in your service. I put port 9080 as first port(which maps to http healthcheck). This way there is no tcp check on broker listener port(in my case port 443)

Here is deployment and service yaml.

```

---
apiVersion: v1
kind: Service
metadata:
  labels:
    app: kafka-proxy-1
  name: kafka-proxy-1
  annotations:
    external-dns.alpha.kubernetes.io/hostname: broker1.mydomain.net
spec:
  ports:
  - name: health
    port: 9080
    protocol: TCP
    targetPort: 9080
  - name: kafka-proxy-1
    port: 443
    protocol: TCP
    targetPort: 443
  selector:
    app: kafka-proxy-1
  sessionAffinity: None
  type: LoadBalancer
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app: kafka-proxy-2
  name: kafka-proxy-2
  annotations:
    external-dns.alpha.kubernetes.io/hostname: broker2.mydomain.net
spec:
  ports:
  - name: health
    port: 9080
    protocol: TCP
    targetPort: 9080
  - name: kafka-proxy-2
    port: 443
    protocol: TCP
    targetPort: 443
  selector:
    app: kafka-proxy-2
  sessionAffinity: None
  type: LoadBalancer
---

apiVersion: v1
kind: Service
metadata:
  labels:
    app: kafka-proxy-3
  name: kafka-proxy-3
  annotations:
    external-dns.alpha.kubernetes.io/hostname: broker3.mydomain.net
spec:
  ports:
  - name: health
    port: 9080
    protocol: TCP
    targetPort: 9080
  - name: kafka-proxy-3
    port: 443
    protocol: TCP
    targetPort: 443
  selector:
    app: kafka-proxy-3
  sessionAffinity: None
  type: LoadBalancer

---
apiVersion: apps/v1
kind: Deployment
metadata:
   name: kafka-proxy-1
spec:
  replicas: 1
  selector:
    matchLabels:
      app: kafka-proxy-1
  template:
    metadata:
      labels:
        app: kafka-proxy-1
    spec:
      hostNetwork: true
      containers:
        - name: kafka-proxy
          securityContext:
            capabilities:
              add: ["NET_ADMIN", "SYS_TIME","NET_BIND_SERVICE"]
          image: grepplabs/kafka-proxy:latest
          args:
            - 'server'
            - '--log-format=json'
            - '--bootstrap-server-mapping=b-1.pdp-kafka.mc3d5e.c6.kafka.us-east-1.amazonaws.com:9094,0.0.0.0:443,broker1.mydomain.net:443'
            - '--external-server-mapping=b-1.pdp-kafka.mc3d5e.c6.kafka.us-east-1.amazonaws.com:9094,broker1.mydomain.net:443'
            - '--external-server-mapping=b-2.pdp-kafka.mc3d5e.c6.kafka.us-east-1.amazonaws.com:9094,broker2.mydomain.net:443'
            - '--external-server-mapping=b-3.pdp-kafka.mc3d5e.c6.kafka.us-east-1.amazonaws.com:9094,broker3.mydomain.net:443'
            - '--tls-enable'
            - '--log-level=debug'
            - '--dynamic-listeners-disable'
            - '--tls-insecure-skip-verify'
            - '--proxy-request-buffer-size=32768'
            - '--proxy-response-buffer-size=32768'
            - '--proxy-listener-read-buffer-size=32768'
            - '--proxy-listener-write-buffer-size=131072'
            - '--kafka-connection-read-buffer-size=131072'
            - '--kafka-connection-write-buffer-size=32768'
            - '--proxy-listener-key-file=/opt/tls/tls.key'
            - '--proxy-listener-cert-file=/opt/tls/tls.crt'
            - '--proxy-listener-tls-enable'
          ports:
          - name: kafka-port1
            containerPort: 443
          - name: health
            containerPort: 9080
          volumeMounts:
          - mountPath: /opt/tls
            name: kafka-wildcard-cert
            readOnly: true
      volumes:
      - name: kafka-wildcard-cert
        secret:
          defaultMode: 420
          secretName: kafka-wildcard-cert

---
apiVersion: apps/v1
kind: Deployment
metadata:
   name: kafka-proxy-2
spec:
  replicas: 1
  selector:
    matchLabels:
      app: kafka-proxy-2
  template:
    metadata:
      labels:
        app: kafka-proxy-2
    spec:
      hostNetwork: true
      containers:
        - name: kafka-proxy
          securityContext:
            capabilities:
              add: ["NET_ADMIN", "SYS_TIME","NET_BIND_SERVICE"]
          image: grepplabs/kafka-proxy:latest
          args:
            - 'server'
            - '--log-format=json'
            - '--bootstrap-server-mapping=b-2.pdp-kafka.mc3d5e.c6.kafka.us-east-1.amazonaws.com:9094,0.0.0.0:443,broker2.mydomain.net:443'
            - '--external-server-mapping=b-1.pdp-kafka.mc3d5e.c6.kafka.us-east-1.amazonaws.com:9094,broker1.mydomain.net:443'
            - '--external-server-mapping=b-2.pdp-kafka.mc3d5e.c6.kafka.us-east-1.amazonaws.com:9094,broker2.mydomain.net:443'
            - '--external-server-mapping=b-3.pdp-kafka.mc3d5e.c6.kafka.us-east-1.amazonaws.com:9094,broker3.mydomain.net:443'
            - '--tls-enable'
            - '--log-level=debug'
            - '--dynamic-listeners-disable'
            - '--tls-insecure-skip-verify'
            - '--proxy-request-buffer-size=32768'
            - '--proxy-response-buffer-size=32768'
            - '--proxy-listener-read-buffer-size=32768'
            - '--proxy-listener-write-buffer-size=131072'
            - '--kafka-connection-read-buffer-size=131072'
            - '--kafka-connection-write-buffer-size=32768'
            - '--proxy-listener-key-file=/opt/tls/tls.key'
            - '--proxy-listener-cert-file=/opt/tls/tls.crt'
            - '--proxy-listener-tls-enable'
          ports:
          - name: kafka-port2
            containerPort: 443
          - name: health
            containerPort: 9080
          volumeMounts:
          - mountPath: /opt/tls
            name: kafka-wildcard-cert
            readOnly: true
      volumes:
      - name: kafka-wildcard-cert
        secret:
          defaultMode: 420
          secretName: kafka-wildcard-cert
---
apiVersion: apps/v1
kind: Deployment
metadata:
   name: kafka-proxy-3
spec:
  replicas: 1
  selector:
    matchLabels:
      app: kafka-proxy-3
  template:
    metadata:
      labels:
        app: kafka-proxy-3
    spec:
      hostNetwork: true
      containers:
        - name: kafka-proxy
          securityContext:
            capabilities:
              add: ["NET_ADMIN", "SYS_TIME","NET_BIND_SERVICE"]
          image: grepplabs/kafka-proxy:latest
          args:
            - 'server'
            - '--log-format=json'
            - '--bootstrap-server-mapping=b-3.pdp-kafka.mc3d5e.c6.kafka.us-east-1.amazonaws.com:9094,0.0.0.0:443,broker3.mydomain.net:443'
            - '--external-server-mapping=b-1.pdp-kafka.mc3d5e.c6.kafka.us-east-1.amazonaws.com:9094,broker1.mydomain.net:443'
            - '--external-server-mapping=b-2.pdp-kafka.mc3d5e.c6.kafka.us-east-1.amazonaws.com:9094,broker2.mydomain.net:443'
            - '--external-server-mapping=b-3.pdp-kafka.mc3d5e.c6.kafka.us-east-1.amazonaws.com:9094,broker3.mydomain.net:443'
            #- '--dial-address-mapping=b-3.pdp-kafka.mc3d5e.c6.kafka.us-east-1.amazonaws.com:9094,0.0.0.0:443'
            - '--tls-enable'
            - '--log-level=debug'
            - '--dynamic-listeners-disable'
            - '--tls-insecure-skip-verify'
            - '--proxy-request-buffer-size=32768'
            - '--proxy-response-buffer-size=32768'
            - '--proxy-listener-read-buffer-size=32768'
            - '--proxy-listener-write-buffer-size=131072'
            - '--kafka-connection-read-buffer-size=131072'
            - '--kafka-connection-write-buffer-size=32768'
            - '--proxy-listener-key-file=/opt/tls/tls.key'
            - '--proxy-listener-cert-file=/opt/tls/tls.crt'
            - '--proxy-listener-tls-enable'
          ports:
          - name: kafka-port3
            containerPort: 443
          - name: health
            containerPort: 9080
          volumeMounts:
          - mountPath: /opt/tls
            name: kafka-wildcard-cert
            readOnly: true
      volumes:
      - name: kafka-wildcard-cert
        secret:
          defaultMode: 420
          secretName: kafka-wildcard-cert

**Comment by srikanthvpai** [comment]
Hi @rajk0007 

Had a question. Why did you disable the dynamic listeners on this ?  Also don't you see Client closed local connection on 172.17.0.2:30002 from 10.223.<<>>.76:60268 (b-2.<<>>.8af4q1.c14.kafka.us-east-1.amazonaws.com:9092)" in the logs of containers ? 

Regards
